### DISCUSSION TITLE
Is it possible to use Queue instead of Topic for certain messages in ASB?

### DISCUSSION THREAD

**Discussion by Arik-Shapiro** [BrighterCommand/Brighter#2243] [Q&A]
I want to be able to use Queues for Commands and Topics for events.
Is it possible to control whether the library uses a queue for a certain messages or a topic?

Thanks.

**Comment by DevJonny** [comment]
Would you be able to provide a little more background as to your use case @Arik-Shapiro?  As a general rule you wouldn't send a command onto a queue except in particular circumstances.

However, as I understand it although generally topics are used to for allowing multiple subscriptions to that queue and all get a copy of that message, there is nothing stopping you only having one subscription to that topic? So the service that needs to receive command can be the only one that subscribes to that topic.

So if we understand a little more about your use case we may be able to advise better, or raise an issue for changes.

**Reply by Arik-Shapiro** [reply]
Well, generally speaking. Commands are usually sent by at least one producer to a single consumer.
On the other hand, an event is sent by one producer and consumed by possibly many consumers.
In the case of a Command I don't see why I'll ever need more than a Queue since there's only one consumer for it.
Of course I can achieve that with a topic because a queue is just a topic with one subscription and no filters.
But I'd love to have the option to just use a queue since in my opinion it is the correct option for a Command.

**Reply by DevJonny** [reply]
Generally speaking, commands aren't sent between services. Yes, and Brighter is built around `.Send(..)` sending a command to exactly one or more handlers, and `.Publish(..)` and `.Post(..)` sending to zero or many handlers (the latter via queues). I agree that you shouldn't ever need to send a Command to more than one queue. However using `.Post(..)` to send commands is an unusual circumstance, which should be considered before proceeding with, hence I was asking what the use case was in your situation. 

However if a Topic + a single subscription really doesn't fit your use case then we can have a look at what options we have. However this would be quite a significant change to how Brighter works for the above mentioned reasons.

@preardon any thoughts?

**Reply by Arik-Shapiro** [reply]
What do you mean commands aren't sent between services? This is exactly how an Orchestration Saga is implemented.
Also Brighter documentation is agreeing with me :)
![image](https://user-images.githubusercontent.com/48131076/186685698-a0d04752-8916-4efc-b192-046d35df13f6.png)
https://www.goparamore.io/building-blocks

A **microservice** acting as a producer can either **send** a command to a queue, that a consumer microservice listens to or...

Also from microservices.io have a look at Orchestration Sagas:
https://microservices.io/patterns/data/saga.html

Again, you could use topics as you've mentioned. But I don't think a topic is a replacement for a queue more like an extension.
A use case for it might be that a topic requires standard tier ASB while queue requires only the basic one.

If it's too much of a hassle to make the change there's no need for it. I was just interested whether the option is available because I didn't see it.

**Reply by DevJonny** [reply]
Apologies, I was being too restrictive in my language. I didn't mean they couldn't be sent, I meant it's unusual that you would. For example in the Saga link you posted they're using a request/reply mechanism which Brighter supports via the `.Call(..)` method. This is one those situations where you would send a command to another service, but that is not as a common scenario as sending an event. As per the Brighter docs you quoted, events are usually preferred over commands:

> The advantage of using an event over a command is that we remove behavioral coupling. The producer has no idea how or if consumers will process the event. See [Smart End points And Dumb Pipes](https://martinfowler.com/articles/microservices.html#SmartEndpointsAndDumbPipes)

Anyway, I'll let @preardon have the final say on whether it would be too much effort to implement this as he's much closer to the ASB implementation in Brighter.

**Reply by preardon** [reply]
At the minute the ASB library does not support Queues, however there is no Reason that it couldn't, we just haven't had a Requirement.

@Arik-Shapiro I would like to hear your requirement.  I completely understand that a command is a 1 to 1 relationship and this fits with the idea of Queues in service Bus however being able to add more subscriptions can be helpful for other reasons, I have seen people putting an "Audit" subscription in place ect. I personally only use Topics in ASB so I have the ability to add subscriptions for other usages.

@DevJonny I do send some of my commands on the bus but this is a scale out tactic and is always within the same Domain. I.e. my pay interest flow is calling an API to trigger chunking of accounts then sending a large number of "Pay Interest to Account" commands. You are correct however the majority of my commands are in proc

**Reply by Arik-Shapiro** [reply]
@preardon 
I work for a company that provides VoIP Recording solutions and we have a product that provides one platform for communications (calls/messages/videos) from multiple integrations like Skype/Teams/Zoom/Session Border Controls(SBC)..

We are in the process of redesigning our system.
I've proposed that each integration will have a service called {Name}Gateway that is responsible for receiving external call messages and transforming it into internal messaging.
The internal messages will be handled by the appropriate bounded context.

For example an integration with SBC which just proxies SIP messages to a service called SignalingGateway.
The SignalingGateway receives SIP messages, they are asynchronous request response messages.
Therefore the most appropriate way to handle them in my opinion is to translate them into internal commands like "StartCall" "StartRecording", "StopRecording" etc..

There are multiple reasons these are commands and not events in my opinion:
1. When receiving a Sip INVITE for example there needs to be a response, doesn't have to be immediate (not synchronous) but a response is expected, for events responses are not usually expected.
2. Integration Events are sent by only one service and consumed by possible many. In this scenario there are multiple gateways emitting the same message that will be handled by the appropriate bounded context.
3. Events are things that "happend" in your system where's a command is an action you want to happen. In this case nothing yet happened.

Just to clarify The flow is something like this:
 SIP Invite -> SignalingGateway ->publish  StartCall (command) -> handled by calls service -> publish CallStarted/FailedToStartCall (from calls service) -> Handle in SignalingGateway -> SIP 200OK
(It's more complex than that there's more commands/events involved)

Now back to why I'm here.
We have a requirement that we might need to publish/consume such messages with multiple different message brokers.
Therefore, I need a lightweight library that provides this kind of abstraction over messaging with many possible transports.
The reason I looked at queues is because we already have used queues in ASB and If I want to easily replace the usage of Azure Service Bus Library with Brighter It needs to support that. Obviously I could just replace them with topics but I'd rather not do that now.

**Reply by preardon** [reply]
@Arik-Shapiro to summarize, you want use queues to integrate with what you already have in your environment?

**Reply by Arik-Shapiro** [reply]
@preardon 
Yes.
Sorry I may have exaggerated with the details 😄. Wasn't sure if you were asking about commands in my case or just why I wanted to use queues for them.

**Reply by preardon** [reply]
@Arik-Shapiro I am more than happy to talk to you about commands / events. I feel what you are describing about is more of an event not a command but that's not the point of your request.

In terms of ASB transport, I'm happy for changes to be made to support Queues as well I think we would need a Bool on ASB subscription to select an the way in as well as one of ASB Public to do the same. Also we would want to ensure intergration tests for queues as well

**Reply by Arik-Shapiro** [reply]
@preardon 
Glad to hear! Should I make a feature request?

**Reply by preardon** [reply]
@Arik-Shapiro Please do, and let me know if you want a chat about design before you get started on the PR :)
